src/ansi-c/library/cprover_contracts.c

```diff
@@ -1418,4 +1421,36 @@ __CPROVER_HIDE:;
 #endif
   }
 }
-#endif
+
+/// \brief Implementation of the `obeys_contract` front-end predicate.
+/// \return True iff a function pointer points to the specified contract.
+///
+/// \details If called in an assumption context, translates to an assignment
+/// `function_pointer = contract`. If called in an assertion context, translates
+/// to an `function_pointer == contract`.
+__CPROVER_bool __CPROVER_contracts_obeys_contract(
+  void (**function_pointer)(void),
+  void (*contract)(void),
+  __CPROVER_contracts_write_set_ptr_t set)
+{
+__CPROVER_HIDE:;
+  __CPROVER_assert(
+    (set != 0) &
+      ((set->assume_requires_ctx == 1) | (set->assert_requires_ctx == 1) |
+       (set->assume_ensures_ctx == 1) | (set->assert_ensures_ctx == 1)),
+    "__CPROVER_obeys_contract is used only in requires or ensures clauses");
+  if((set->assume_requires_ctx == 1) | (set->assume_ensures_ctx == 1))
+  {
+    if(*function_pointer != 0)
+    {
+      *function_pointer = contract;
```

Comment:
So you are assigning only when the pointer is one towards an actual function, and not a null pointer? Doing a null-check and then assigning without an additional level of dereferencing still looks weird. This might warrant a comment.

Reply:
I'll revert to an unguarded assignment and let the user order null checks and predicate calls in contracts instead.